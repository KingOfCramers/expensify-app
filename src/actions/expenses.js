import uuid from "uuid";
import moment from "moment";

// Action Generators
  // ADD_EXPENSE action generator
   export const addExpense = ({ description = "", note = "", amount = 0, createdAt = moment().valueOf() } = {}) => ({
      type: "ADD_EXPENSE",
      expense: {
        id: uuid(), // uuid is temporary alternative for id, will come from mongodb later
        description,
        note,
        amount,
        createdAt
      }
    })
  // REMOVE_EXPENSE action generator
   export const removeExpense = ({ id } = {}) => ({
      type: "REMOVE_EXPENSE",
      id
    })
  // EDIT_EXPENSE action generator
   export const editExpense = (id, updates) => ({
      type: "EDIT_EXPENSE",
      id,
      updates
    })