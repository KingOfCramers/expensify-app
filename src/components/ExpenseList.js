import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";
// Unconnected //
export const ExpenseList = (props) => ( // Exporting for testing...
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {props.expenses.length === 0 ? (
        <div className="list-item--empty">
          <p>No expenses</p>
        </div>
      ) : (
        props.expenses.map((expense) => {
          return <ExpenseListItem key={expense.id} {...expense}/>
        })
      )}
    </div>
  </div>
);

const mapStateToProps = (state) => { // This lets us determine what aspects of the redux state we want to pass in.
  return {
    expenses: selectExpenses(state.expenses, state.filters) // Use the values on our state object to calculate the visible expenses.
  }
}

export default connect(mapStateToProps)(ExpenseList); // "connect" is a higher order component.