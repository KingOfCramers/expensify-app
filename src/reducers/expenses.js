const expensesReducerDefaultState = []; // Empty array...
const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch(action.type) {
    case "ADD_EXPENSE" :
      return [...state, action.expense ]; // Don't use push!
    case "REMOVE_EXPENSE" :
      return state.filter((expense) => {
        return expense.id !== action.expense.id;
      });
    case "EDIT_EXPENSE" :
      return state.map((expense) => {
        if(expense.id === action.id){
          return {
            ...expense,
            ...action.updates // Override any properties from update
          };
        } else {
          return expense;
        }
      })
    default:
      return state;
  }
};

export default expensesReducer;