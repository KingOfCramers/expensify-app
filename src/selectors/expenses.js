import moment from "moment";

// Get visible expenses (takes in current Redux state.expenses and state.filters)
// Returns an array of visible expenses...
  const selectExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
      const createdAtMoment = moment(expense.createdAt); // .createdAt is raw number, will output moment object w/ correct date.
      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true; // if startDate filter prop exists, check to see if it (it's a moment) is same or before the createdAt prop from this expense. If it is, return false (don't filter out expense)
      const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true;  // Same as above...
      const textMatch = expense.description.toLowerCase().includes(text.toLowerCase()) || expense.note.toLowerCase().includes(text.toLowerCase()) // If an empty string, return true (don't filter) OR if note/description includes text, don't filter (no need for text === '' condition, all strings will have it)
      return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
      if(sortBy === "date"){
        return a.createdAt > b.createdAt ? -1 : 1; // If "a" is more recent, return -1 (which indexes "a" first), otherwise, index "b" first.
      } else if (sortBy === "amount"){
        return a.amount > b.amount ? -1 : 1;
      } else {
        return 0; // Else, do not sort.
      }
    });
  }

export default selectExpenses;