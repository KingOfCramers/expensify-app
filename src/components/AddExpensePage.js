import React from 'react';
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense); // Passing this in as a prop (using the mapdispatch to props ) to allow for easier testing, rather than using global dispatch function.
    this.props.history.push("/")
  }
  render(){
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">Add Expense</h2>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
};

const mapDispatchToProps = (dispatch) => { // We have access to dispatch thanks to react-redux. The goal is to simplify testing, by
  return {
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
  }
};
export default connect(undefined, mapDispatchToProps)(AddExpensePage); // Second argument is matchDispatchToProps