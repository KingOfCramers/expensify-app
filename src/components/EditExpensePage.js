import React from 'react';
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, startRemoveExpense } from "../actions/expenses";

/// GIT LEARNING TO ROLLBACK
export class EditExpensePage extends React.Component {

  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push("/")
  };

  onRemove = () => {
    this.props.startRemoveExpense({id: this.props.expense.id }) // .then(() => { console.log("This worked!")} ) // You could chain on an action here!
    this.props.history.push("/")
  };
 //
  render(){
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => { // we have access to the props as the second argument here.
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id ) // the props.match.params comes from the Router (it's part of the URL)
  }
};

const mapDispatchToProps = (dispatch, props) => {
  return {
    startRemoveExpense: (expense) => { return dispatch(startRemoveExpense(expense))}, // This returns a promise (the bottomn line is implicit return)... See the promise return example above.
    editExpense: (id, expense) => dispatch(editExpense(id, expense))
  }
}


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);