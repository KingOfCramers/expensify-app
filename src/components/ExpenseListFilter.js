import { connect } from "react-redux";
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from "../actions/filters";
import { DateRangePicker } from "react-dates";
import React from "react";

export class ExpenseListFilter extends React.Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }))
  };

  onSortChange = (e) => {
    if(e.target.value === "date"){
      this.props.sortByDate();
    } else {
      this.props.sortByAmount();
    }
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value)
  };

  render(){
    return (
      <div>
          <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
          <select onChange={this.onSortChange}>
            <option value="date">Date</option>
            <option value="amount">Amount</option>
          </select>
          <DateRangePicker
            startDate={this.props.filters.startDate} // Accepts moment object
            startDateId={"fluffykins"}
            endDateId={"fluffykins"}
            endDate={this.props.filters.endDate} // Accepts moment object
            onDatesChange={this.onDatesChange} // passes startDate and endDate to function, to set new state.
            focusedInput={this.state.calendarFocused}
            onFocusChange={this.onFocusChange} // Sets focus state to calendarFocused state.
            showClearDates={true}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
        </div>
    );
  }
}

const mapStateToProps = (state) => { // This lets us determine what aspects of the redux state we want to pass in.
  return {
    filters: state.filters
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate))
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);