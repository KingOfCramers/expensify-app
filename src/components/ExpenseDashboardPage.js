import React from 'react';
import ExpenseList from "./ExpenseList";
import ExpenseListFilter from "./ExpenseListFilter";
import ExpenseSummary from "./ExpenseSummary";
import { connect } from "react-redux";

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseSummary />
    <ExpenseListFilter />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;