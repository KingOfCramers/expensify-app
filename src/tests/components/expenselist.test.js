import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

test("should render ExpenseList w/ expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />)
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseList w/ empty message", () => {
  const wrapper = shallow(<ExpenseList expenses={[]}/>);
  expect(wrapper).toMatchSnapshot();
});