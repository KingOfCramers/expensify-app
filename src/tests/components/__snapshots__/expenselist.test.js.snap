// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList w/ empty message 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item--empty"
    >
      <p>
        No expenses
      </p>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseList w/ expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseListItem
      amount={1000}
      createdAt={-345600000}
      description="Sledge 1"
      id="1"
      key="1"
      note=""
    />
    <ExpenseListItem
      amount={3000}
      createdAt={0}
      description="Creep 2"
      id="2"
      key="2"
      note=""
    />
    <ExpenseListItem
      amount={2000}
      createdAt={345600000}
      description="Zombat 3"
      id="3"
      key="3"
      note=""
    />
  </div>
</div>
`;
